File: backend/node-express/src/routes/question.routes.js
const express = require('express');
const router = express.Router();
const questionController = require('../controllers/question.controller');
const { createQuestionDto, updateQuestionDto, searchQuestionDto, paginatedQuestionDto } = require('../dto/question.dto');
const { validationResult } = require('express-validator');

// Create Question
router.post('/', createQuestionDto, (req, res, next) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return res.status(400).json({ success: false, errors: errors.array() });
  }
  questionController.create(req, res, next);
});

// Get All Questions
router.get('/', questionController.getAll.bind(questionController));

// Get Questions Paginated
router.get('/paginated', paginatedQuestionDto, (req, res, next) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return res.status(400).json({ success: false, errors: errors.array() });
  }
  questionController.getPaginated(req, res, next);
});

// Search Questions
router.get('/search', searchQuestionDto, (req, res, next) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return res.status(400).json({ success: false, errors: errors.array() });
  }
  questionController.search(req, res, next);
});

// Get Hot Questions
router.get('/hot', questionController.getHot.bind(questionController));

// Get Question By ID
router.get('/:id', questionController.getById.bind(questionController));

// Get Related Questions
router.get('/:id/related', questionController.getRelated.bind(questionController));

// Update Question
router.put('/:id', updateQuestionDto, (req, res, next) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return res.status(400).json({ success: false, errors: errors.array() });
  }
  questionController.update(req, res, next);
});

// Delete Question
router.delete('/:id', questionController.delete.bind(questionController));

module.exports = router;

File: backend/python-fastapi/routes/question.routes.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import List
from schemas.schemas import QuestionCreate, QuestionUpdate, QuestionResponse, PaginatedQuestions
from services.question.service import QuestionService
from config.database import get_db
from jose import jwt, JWTError
from dotenv import load_dotenv
import os

load_dotenv()
SECRET_KEY = os.getenv("SECRET_KEY", "your-secret-key")
ALGORITHM = os.getenv("ALGORITHM", "HS256")

router = APIRouter()


def get_current_user(authorization: str, db: Session = Depends(get_db)):
    try:
        if not authorization.startswith("Bearer "):
            raise HTTPException(status_code=401, detail="Invalid authorization header")
        token = authorization.split(" ")[1]
        payload = jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
        return payload
    except JWTError:
        raise HTTPException(status_code=401, detail="Invalid token")


@router.post("/", response_model=QuestionResponse)
def create_question(
    request: QuestionCreate,
    authorization: str = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    question = QuestionService.create(
        db, request.title, request.description, request.status,
        authorization["userId"], authorization["username"]
    )
    return question


@router.get("/", response_model=List[QuestionResponse])
def get_all_questions(db: Session = Depends(get_db)):
    return QuestionService.get_all(db)


@router.get("/paginated", response_model=PaginatedQuestions)
def get_paginated_questions(
    page: int = Query(1, ge=1),
    limit: int = Query(10, ge=1, le=100),
    db: Session = Depends(get_db)
):
    return QuestionService.get_paginated(db, page, limit)


@router.get("/search", response_model=List[QuestionResponse])
def search_questions(q: str = Query(...), db: Session = Depends(get_db)):
    return QuestionService.search(db, q)


@router.get("/hot", response_model=List[QuestionResponse])
def get_hot_questions(limit: int = Query(5, ge=1), db: Session = Depends(get_db)):
    return QuestionService.get_hot(db, limit)


@router.get("/{id}", response_model=QuestionResponse)
def get_question_by_id(id: str, db: Session = Depends(get_db)):
    return QuestionService.get_by_id(db, id)


@router.get("/{id}/related", response_model=List[QuestionResponse])
def get_related_questions(id: str, limit: int = Query(5, ge=1), db: Session = Depends(get_db)):
    return QuestionService.get_related(db, id, limit)


@router.put("/{id}", response_model=QuestionResponse)
def update_question(
    id: str,
    request: QuestionUpdate,
    authorization: str = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    return QuestionService.update(
        db, id, request.title, request.description, request.status,
        authorization["userId"]
    )


@router.delete("/{id}")
def delete_question(
    id: str,
    authorization: str = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    QuestionService.delete(db, id, authorization["userId"])
    return {"success": True, "message": "Question deleted successfully"}
